cmd/thanos/receive.go

```diff
@@ -214,235 +214,302 @@ func runReceive(
-					level.Error(logger).Log("err", err, "msg", "failed to flush storage")
-				} else {
-					level.Info(logger).Log("msg", "storage is flushed successfully")
-				}
-				if err := dbs.Close(); err != nil {
-					level.Error(logger).Log("err", err, "msg", "failed to close storage")
-					return
-				}
-				level.Info(logger).Log("msg", "storage is closed")
-			}()
+			statusProber.Healthy()
 
-			for {
-				select {
-				case <-cancel:
-					return nil
-				case _, ok := <-hashringChangedChan:
-					if !ok {
-						return nil
-					}
-					dbUpdatesStarted.Inc()
-					level.Info(logger).Log("msg", "updating storage")
-
-					if err := dbs.Flush(); err != nil {
-						return errors.Wrap(err, "flushing storage")
-					}
-					if err := dbs.Open(); err != nil {
-						return errors.Wrap(err, "opening storage")
-					}
-					if upload {
-						uploadC <- struct{}{}
-						<-uploadDone
-					}
-					statusProber.Ready()
-					level.Info(logger).Log("msg", "storage started, and server is ready to receive web requests")
-					dbUpdatesCompleted.Inc()
-					dbReady <- struct{}{}
-				}
-			}
+			return srv.ListenAndServe()
 		}, func(err error) {
-			close(cancel)
+			statusProber.NotReady(err)
+			defer statusProber.NotHealthy(err)
+
+			srv.Shutdown(err)
 		})
 	}
 
-	level.Debug(logger).Log("msg", "setting up hashring")
+	level.Debug(logger).Log("msg", "setting up grpc server")
 	{
-		// Note: the hashring configuration watcher
-		// is the sender and thus closes the chan.
-		// In the single-node case, which has no configuration
-		// watcher, we close the chan ourselves.
-		updates := make(chan receive.Hashring, 1)
-
-		// The Hashrings config file path is given initializing config watcher.
-		if conf.hashringsFilePath != "" {
-			cw, err := receive.NewConfigWatcher(log.With(logger, "component", "config-watcher"), reg, conf.hashringsFilePath, *conf.refreshInterval)
-			if err != nil {
-				return errors.Wrap(err, "failed to initialize config watcher")
+		var s *grpcserver.Server
+		startGRPC := make(chan struct{})
+
+		if err := setupGRPCServer(g, logger, reg, tracer, conf, s, startGRPC, dbReady, comp, dbs, webHandler, grpcLogOpts, tagOpts, grpcProbe); err != nil {
+			return err
+		}
+	}
+
+	level.Debug(logger).Log("msg", "setting up receive http handler")
+	{
+		g.Add(
+			func() error {
+				return errors.Wrap(webHandler.Run(), "error starting web server")
+			},
+			func(err error) {
+				webHandler.Close()
+			},
+		)
+	}
+
+	level.Info(logger).Log("msg", "starting receiver")
+	return nil
+}
+
+func setupGRPCServer(g *run.Group,
+	logger log.Logger,
+	reg *prometheus.Registry,
+	tracer opentracing.Tracer,
+	conf *receiveConfig, s *grpcserver.Server,
+	startGRPC chan struct{},
+	dbReady chan struct{},
+	comp component.SourceStoreAPI,
+	dbs *receive.MultiTSDB,
+	webHandler *receive.Handler,
+	grpcLogOpts []grpc_logging.Option,
+	tagOpts []tags.Option,
+	grpcProbe *prober.GRPCProbe,
+
+) error {
+	g.Add(func() error {
```

Comment:
This looks it does the job. I'd prefer keeping the `run.Group` in the top-level function and extract whatever we pass to `g.Add` on a different function.